documentation/bolt_types_reference.md

```diff
@@ -0,0 +1,319 @@
+
+**Returns**
+
+`ResultSet`: The set of failing results.
+
+### `filter_set(block)`
+
+**Parameters**
+
+- `block`
+
+  The block to filter the set by.
+
+**Returns**
+
+`ResultSet`: The set of filtered results.
+
+### `find(String $target_name)`
+
+**Parameters**
+
+- `String` target_name
+
+  The string name of the target to retrive a result for.
+
+**Returns**
+
+`Variant[Result, ApplyResult]`: The target's result.
+
+### `first`
+
+**Returns**
+
+`Variant[Result, ApplyResult]`: The first result in the set. Useful for unwrapping single results.
+
+### `names`
+
+**Returns**
+
+`Array[String]`: The names of all targets that have results in the set.
+
+### `ok`
+
+**Returns**
+
+`Boolean`: Whether all results were successful. Equivalent to `$result_set.error_set.empty`.
+
+### `ok_set`
+
+**Returns**
+
+`ResultSet`: The set of successful results.
+
+### `results`
+
+**Returns**
+
+`Array[Variant[Result, ApplyResult]]`: All results in the set.
+
+### `targets`
+
+**Returns**
+
+`Array[Target]`: The list of targets that have results in the set.
+
+### `to_data`
+
+**Returns**
+
+`Array[Hash`: An array of serialized representations of each result in the set.
+
+
+## `Target`
+
+The `Target` object represents a target and its specific connection options.
+
+The following functions are available to `ResultSet` objects. They are accessed with
+dot notation `$result_set.function`.
+
+### `config`
+
+**Returns**
+
+`Hash[String, Data]`: The target-level configuration.
+
+> **Note:** This function does not return default configuration values or
+  configuration set in a `bolt.yaml` file.
+
+### `facts`
+
+**Returns**
+
+`Hash[String, Data]`: The target's facts.
+
+> **Note::** This function does not lookup facts for a target and only returns
+  the facts specified in an `inventory.yaml` file or set on a target during a
+  plan run.
+
+### `features`
+
+**Returns**
+
+`Array[String]`: The target's features.
+
+### `host`
+
+**Returns**
+
+`String`: The target's hostname.
+
+### `name`
+
+**Returns**
+
+`String`: The target's human-readable name, or its URI if a name was not given.
+
+### `password`
+
+**Returns**
+
+`String`: The password to use when connecting to the target.
+
+### `plugin_hooks`
+
+**Returns**
+
+`Hash[String, Data]`: The target's 
+[`plugin_hooks` configuration options](https://puppet.com/docs/bolt/latest/bolt_configuration_reference.html#plugin-hooks-configuration-options)
+
+### `port`
+
+**Returns**
+
+`Integer`: The target's connection port.
+
+### `protocol`
+
+**Returns**
+
+`String`: The protocol used to connect to the target.
+
+### `safe_name`
+
+**Returns**
+
+`String`: The target's safe name. Equivalent to `name` if a name was given, or
+the target's `uri` with any password omitted.
+
+### `target_alias`
+
+**Returns**
+
+`Variant[String, Array[String]]`: The target's aliases.
+
+### `uri`
+
+**Returns**
+
+`String`: The target's URI.
+
+### `user`
+
+**Returns**
+
+`String`: The user to run as on the target.
+
+### `vars`
```

Comment:
https://puppet.com/docs/puppet/latest/lang_data.html I don't see the `Error` datatype documented in Puppet, but we should probably find a way to document it and include it since we reference it. Ideally adding it to the Puppet docs and linking would be the way to go, but if that's going to take a while including here might not be a bad idea?

Reply:
I'll open a ticket for this bit and add it to the Puppet docs